### PR TITLE
feat: Validate if a BuildItem is final or abstract

### LINKED ISSUE #926: Validate if a BuildItem is `final` or `abstract`
See 

 * IJ PR https://github.com/redhat-developer/intellij-quarkus/pull/1171

BuildItems consumed or produced in BuildSteps must be final or abstract. It would be nice if the IDE could warn about that when creating a new BuildItem subclass.

Eg. 
`public abstract class AbstractDevUIBuildItem extends MultiBuildItem {...}`
`public final class QuteTemplateBuildItem extends AbstractDevUIBuildItem {...}`

initially reported as https://github.com/redhat-developer/intellij-quarkus/issues/1120.

Exemples of BuildItems: https://github.com/search?q=repo%3Aquarkiverse%2Fquarkus-ironjacamar%20BuildItem&type=code

### PR DESCRIPTION
**Pull request by angelozerr**
feat: Validate if a BuildItem is final or abstract

Fixes #926

Here the result:

image

### CODE REVIEW

File: quarkus.jdt.ext/com.redhat.microprofile.jdt.quarkus/src/main/java/com/redhat/microprofile/jdt/internal/quarkus/builditems/java/QuarkusBuildItemASTVisitor.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+		}
+		return false;
+	}
+
+	private void validateBuildItem(TypeDeclaration typeDeclaration, JavaDiagnosticsContext context) {
+		if (isValidBuildItem(typeDeclaration)) {
+			return;
+		}
+		super.addDiagnostic(createDiagnosticMessage(typeDeclaration, context.getDocumentFormat()),
+				QuarkusConstants.QUARKUS_DIAGNOSTIC_SOURCE, typeDeclaration,
```
Comment (datho7561) [inline quarkus.jdt.ext/com.redhat.microprofile.jdt.quarkus/src/main/java/com/redhat/microprofile/jdt/internal/quarkus/builditems/java/QuarkusBuildItemASTVisitor.java:77]: This highlights the entire class body; I think it makes more sense to highlight just the name of the type.

**Review by datho7561 — CHANGES_REQUESTED** [review]

File: quarkus.jdt.ext/com.redhat.microprofile.jdt.quarkus/src/main/java/com/redhat/microprofile/jdt/internal/quarkus/builditems/java/QuarkusBuildItemASTVisitor.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+	/**
+	 * Returns true if the given type declaration extends
+	 * 'io.quarkus.builder.item.BuildItem' class and false otherwise.
+	 * 
+	 * @param typeDeclaration the type declaration.
+	 * @return true if the given type declaration extends
+	 *         'io.quarkus.builder.item.BuildItem' class and false otherwise.
+	 */
+	private static boolean isBuildItem(TypeDeclaration typeDeclaration) {
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply quarkus.jdt.ext/com.redhat.microprofile.jdt.quarkus/src/main/java/com/redhat/microprofile/jdt/internal/quarkus/builditems/java/QuarkusBuildItemASTVisitor.java:64]: Thanks!

File: quarkus.jdt.ext/com.redhat.microprofile.jdt.quarkus/src/main/java/com/redhat/microprofile/jdt/internal/quarkus/builditems/java/QuarkusBuildItemASTVisitor.java (hunk 1 above)
Comment (angelozerr) [reply quarkus.jdt.ext/com.redhat.microprofile.jdt.quarkus/src/main/java/com/redhat/microprofile/jdt/internal/quarkus/builditems/java/QuarkusBuildItemASTVisitor.java:77]: Indeed, I fixed that:

image

File: quarkus.jdt.ext/com.redhat.microprofile.jdt.quarkus/src/main/java/com/redhat/microprofile/jdt/internal/quarkus/builditems/java/QuarkusBuildItemASTVisitor.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+			return;
+		}
+		super.addDiagnostic(createDiagnosticMessage(typeDeclaration, context.getDocumentFormat()),
+				QuarkusConstants.QUARKUS_DIAGNOSTIC_SOURCE, typeDeclaration,
+				QuarkusBuildItemErrorCode.InvalidModifierBuildItem, DiagnosticSeverity.Error);
+	}
+
+	private static boolean isValidBuildItem(TypeDeclaration typeDeclaration) {
+		int flags = typeDeclaration.getModifiers();
+		return Modifier.isStatic(flags) || Modifier.isFinal(flags);
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply quarkus.jdt.ext/com.redhat.microprofile.jdt.quarkus/src/main/java/com/redhat/microprofile/jdt/internal/quarkus/builditems/java/QuarkusBuildItemASTVisitor.java:83]: Good catch! fixed.

[the remark this answers is not in the document]
Comment (datho7561) [reply quarkus.jdt.ext/com.redhat.microprofile.jdt.quarkus/src/main/java/com/redhat/microprofile/jdt/internal/quarkus/builditems/java/QuarkusBuildItemASTVisitor.java:83]: It hasn't changed

**Comment by angelozerr** [comment]
> Seems good to me!

Thanks @datho7561 for your review!

**Pull request merged**